`CONTRIBUTING.md` extended with section regarding unit tests (#16372)
status-im/status-desktop: CONTRIBUTING.md

  relative import from the file system. As a consequence, the mechanism for
  overriding import paths for tests and Storybook's pages will not work.

## Unit tests

The project has unit tests for both C++ and QML code.

### C++ unit tests

> C++ unit tests are localized under `ui/StatusQ/tests`. They can be executed

Reviewer: ```suggestion
C++ unit tests are located under `ui/StatusQ/tests`. They can be executed
```

> C++ unit tests are localized under `ui/StatusQ/tests`. They can be executed
> easily from QtCreator. Just open the `StatusQ` project in `QtCreator`
> (`status-desktop/ui/StatusQ/CMakeLists.txt`) and select the appropriate target (each test is a separate target).

Reviewer 2: ```suggestion
C++ unit tests are localized under `ui/StatusQ/tests`. They can be executed
easily from QtCreator or make target:

- Open the `StatusQ` project in `QtCreator` (`status-desktop/ui/StatusQ/CMakeLists.txt`) and select the appropriate target (each test is a separate target).

- Run the `run-statusq-tests` target. E.g. `make -j10 run-statusq-tests ARGS="-R ModelEntryTest"`

```

Reviewer 2: not sure QtCreator is a popular IDE around contributors.

Author: I'm not sure either. But it's good to add at least some hints for QtCreator which is used commonly by CCs.

Reviewer 2: That's what I was trying to point out. I'm not sure it's commonly used by CCs 😄 
AFAIK most people use it only when they don't know of any other alternative to accomplish something (like running tests).

### QML unit tests

> QML tests are localized under `storybook/qmlTests/tests` and `ui/StatusQ/tests`. The

Reviewer: ```suggestion
QML tests are located under `storybook/qmlTests/tests` and `ui/StatusQ/tests`. The
```

first location is recommended for all new tests because it provides integration with
the `Storybook`.

Unit tests operate without backend. Backend layer is cut-off using stubs mechanism in the same
way as for pages in `Storybook`.

> QML tests can be executed by running `QmlTests` target of `Storybook` project

Reviewer 2: just realised there's no option of running a specific test using the make target. We're missing the `ARGS` option. I'm usually using something like `storybook/build/bin/QmlTests -input /Users/alexjbanca/Repos/status-desktop/storybook/qmlTests/tests/tst_ConnectDAppModal.qml` but won't work without building the tests before. Maybe something to be easily added for contributors that are not using QtCreator.

Author: It's not enough to add `${ARGS}` because the test is run via `ctest` and it seem that there is no simple way to pass sth to the actual test...

  additional component introduces a dependency on it in the test, even if the component under test
  itself is independent of it.

> [!NOTE]
> Separately testing components that are used together in an application may lead to the
> conclusion that it is uncertain whether they are compatible. However, the point is not to
> not test it, but to do it in an appropriate test (usually in test of component using those
> two components together).

- Unit tests should test units.

  This statement is also true regarding subcomponents used in a component tested in a given
  unit test. Subcomponents should not be subject of testing in unit test (as they form
  a different unit), they should be subject of testing in their own unit test. The unit tests
  should verify if subcomponents are used in the intended way and simply assume that
  internally they work as expected.

  Respecting this rule will help keep unit tests smaller, better structured and easier
  to maintain.

- Unit tests should execute very quickly.

  Relying on `wait(...)` calls slows down tests and is unreliable, may lead to flakey tests,
  therefore should be avoided.

> - Unit tests should not generate any warnings and debug logs.

Reviewer 2: I'm not sure what's the intent here. Is it referring that the test itself should not generate warnings and debug logs, or that there should be no warnings/debug logs generated by the component under test?

Totally agree with the first one, not sure about the second one 😄

Author: I think both. I added note about expected warnings from the component.

Reviewer 2: I think it's good to have a goal of warning free tests, but probably not a rule.

It's quite difficult to achieve a warning free test because even some low level components will generate warnings. Sometimes it depends on Qt version. A good example here is the `StatusButton`. IIRC it shows a polish loop warning on some qt versions or platforms.
